start.S

```diff
@@ -0,0 +1,62 @@
+/* With minor changes, this start code was taken from:
+ * https://github.com/bztsrc/raspi3-tutorial
+ *
+ * Thank you!
+ */
+
+.section ".text.boot"
+
+.global _start
+
+_start:
+	/* Read Multiprocessor Affinity Register with special MRS instruction
+	 * used to access system registers to identify what core we're running on.
+	 * Reference: https://developer.arm.com/docs/ddi0500/g/system-control/aarch64-register-descriptions/multiprocessor-affinity-register
+	 */
+	mrs     x1, mpidr_el1
+
+	/* Cortex-A53 is equipped with 4 cores (all of which are running), so
+	 * we will want to park the other 3 while the primary core (core 0)
+	 * performs set-up.
+	 *
+	 * This is a 64-bit register where the core identifier is in the last
+	 * 2 bits (4 cores, only 2 bits needed to represent IDs up to 4). Mask
+	 * them off. If the identifier is 0, then this is the primary core with
+	 * which we'll perform set-up with.
+	 */
+	and     x1, x1, #0x3
+	cbz     x1, begin_setup
+wait_here:
+	/* Otherwise, we'll park the core here and tell it to wait for an event.
+	 * (presumably from us)
+	 * Reference: https://developer.arm.com/docs/den0024/latest/the-a64-instruction-set/system-control-and-other-instructions/hint-instructions
+	 */
+	wfe
+	b 	wait_here
+
+begin_setup:
+	/* Move the stack to grow down from where our code begins. */
+	ldr     x1, =_start
+	mov     sp, x1
+
+	/* Load the bss (refer to linker script to see where these are) */
+	ldr     x1, =__bss_start
+	ldr     w2, =__bss_size
+
+clear_bss:
+	/* Global variables live in the bss section, and therefore we must zero it
+	 * out.
+	 *
+	 * We've stored the size of the bss in w2, so once this hits zero, we
+	 * have cleared the entire section.
+	 */
+	cbz     w2, enter_kernel
+	str     xzr, [x1], #8
```

Comment:
Does this work or is this just writing a zero to the same place for this entire loop?

Reply:
I think you're right and this just writes zero to the same place, i think if we swapped the `sub` and `str` and changed the `str` to be `str     xzr, [x1], w2` or something like that it'll do the right thing

Reply:
https://github.com/bztsrc/raspi3-tutorial/issues/58#issuecomment-538686629